src/parquet/ReadParquet.cpp

```diff
@@ -1,5 +1,407 @@
+
+// TODO The implementation seemingly reads 128-bit decimal values into
+// Chapel `real`s (or C `double`s). I don't see how this can work properly with
+// large enough data. For now, I think this keeps the implementation practically
+// same as the old, likely carrying the same bug into the new implementation,
+// unfortunately.
+template<> struct ChplBuf<parquet::FLBAType>  {using ElemType = double;};
+
+// this struct is used to bundle different types into one, based on a single
+// ArrowType
+template<typename ArrowType>
+struct TypeBundle {
+  using ReaderType = typename parquet::TypedColumnReader<ArrowType>;
+  using ChplType = typename ChplBuf<ArrowType>::ElemType;
+  using PqType = typename ArrowType::c_type;
+};
+
+template<>
+struct TypeBundle<arrow::Decimal128Type> {
+  using ReaderType = typename parquet::TypedColumnReader<parquet::FLBAType>;
+  using ChplType = typename ChplBuf<parquet::FLBAType>::ElemType;
+  using PqType = typename parquet::FixedLenByteArray;
+};
+} // end anonymous namespace
+
+// Engin: We could consider exposing this struct to Chapel. Note that in
+// ParquetMsg.chpl there is a somewhat similar type. However, this one is for a
+// single column read
+// Based on what we decide, the implementation of this struct can be mode more
+// C++-like or could stay more C-like.
+struct AkColReadOp {
+  void* chpl_arr;
+  int64_t *startIdx;
+  std::shared_ptr<parquet::ColumnReader> column_reader;
+  bool hasNonFloatNulls;
+  chplEnum_t nullMode;
+  int64_t row_idx;
+  int64_t numElems;
+  int64_t batchSize;
+  bool* where_null_chpl;
+  const parquet::ColumnDescriptor* col_info;
+
+  // Generic read implementation
+  template<typename ArrowType>
+  int64_t read() {
+    using Types = TypeBundle<ArrowType>;
+    using ChplType = typename Types::ChplType;
+    using ReaderType = typename Types::ReaderType;
+
+    auto chpl_ptr = (ChplType*)chpl_arr;
+    int64_t num_read = 0;
+    ReaderType* reader = static_cast<ReaderType*>(column_reader.get());
+    *startIdx -= reader->Skip(*startIdx);
+
+    // TODO find better variable names. values_read and num_read are confusing
+    int64_t values_read = 0;
+
+    // note that floats are the types that natively support nulls. This generic
+    // read() shouldn't do anything with floats
+    if (nullMode == noNulls || nullMode == onlyFloats) {
+      while (reader->HasNext() && row_idx < numElems) {
+        if((numElems - row_idx) < batchSize) // adjust batchSize if needed
+          batchSize = numElems - row_idx;
+        reader->ReadBatch(batchSize, nullptr, nullptr, &chpl_ptr[row_idx],
+                          &values_read);
+        row_idx+=values_read;
+        num_read+=values_read;
+      }
+    }
+    else {
+      int16_t definition_level; // nullable type and only reading single records in batch
+      while (reader->HasNext() && row_idx < numElems) {
+        (void)reader->ReadBatch(1, &definition_level, nullptr,
```

Comment:
In the above block, you have `reader->ReadBatch` without the cast to void. Why is it needed here?

I assume its to silence unused return warnings? If so, please add the cast to void above

Reply:
Jade taught me about `std::ignore`. So, I used that in cases like this where I introduced this unused return values.